File: ts/src/storage/files.ts
// File asset storage — store, serve, cleanup, MIME validation

import type { FileRef } from "../types.ts";

// Magic bytes for common MIME types
const MAGIC_BYTES: Record<string, number[]> = {
  "image/png": [0x89, 0x50, 0x4e, 0x47],
  "image/jpeg": [0xff, 0xd8, 0xff],
  "image/gif": [0x47, 0x49, 0x46],
  "image/webp": [0x52, 0x49, 0x46, 0x46], // RIFF (first 4 bytes)
  "application/pdf": [0x25, 0x50, 0x44, 0x46], // %PDF
};

// File extension to MIME type mapping
const EXT_TO_MIME: Record<string, string> = {
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".pdf": "application/pdf",
  ".json": "application/json",
  ".txt": "text/plain",
  ".html": "text/html",
  ".css": "text/css",
  ".js": "application/javascript",
  ".svg": "image/svg+xml",
  ".mp4": "video/mp4",
  ".mp3": "audio/mpeg",
  ".wav": "audio/wav",
  ".zip": "application/zip",
};

export function mimeFromExtension(filename: string): string {
  const ext = filename.substring(filename.lastIndexOf(".")).toLowerCase();
  return EXT_TO_MIME[ext] ?? "application/octet-stream";
}

export function validateMimeType(
  declaredMime: string,
  allowedMimeTypes: string[],
): boolean {
  if (allowedMimeTypes.length === 0) return true;
  return allowedMimeTypes.some((allowed) => {
    if (allowed === declaredMime) return true;
    if (allowed.endsWith("/*")) {
      return declaredMime.startsWith(allowed.slice(0, -1));
    }
    return false;
  });
}

export function validateMagicBytes(data: Uint8Array, declaredMime: string): boolean {
  const expected = MAGIC_BYTES[declaredMime];
  if (!expected) return true; // No magic bytes check for unknown types
  if (data.byteLength < expected.length) return false;
  return expected.every((b, i) => data[i] === b);
}

export function sanitizeFilename(name: string): string {
  // Remove path separators and dangerous chars
  return name
    .replace(/[/\\:*?"<>|]/g, "_")
    .replace(/\.\./g, "_")
    .trim() || "unnamed";
}

function hashFilename(data: Uint8Array, originalName: string): string {
  // Simple FNV-1a hash of file contents for unique naming
  let h = 0x811c9dc5;
  for (let i = 0; i < data.byteLength; i++) {
    h ^= data[i];
    h = Math.imul(h, 0x01000193);
  }
  const hash = (h >>> 0).toString(36);
  const ext = originalName.includes(".") ? originalName.substring(originalName.lastIndexOf(".")).toLowerCase() : "";
  return hash + ext;
}

export async function storeFile(
  dataDir: string,
  tableName: string,
  rowId: string,
  fieldName: string,
  filename: string,
  data: Uint8Array,
  mime: string,
): Promise<FileRef> {
  const hashedName = hashFilename(data, filename);
  const dirPath = `${dataDir}/_files/${tableName}/${rowId}/${fieldName}`;
  await Deno.mkdir(dirPath, { recursive: true });

  const filePath = `${dirPath}/${hashedName}`;
  await Deno.writeFile(filePath, data);

  const relativePath = `_files/${tableName}/${rowId}/${fieldName}/${hashedName}`;
  return {
    path: relativePath,
    name: filename,
    size: data.byteLength,
    mime,
    url: `/api/files/${tableName}/${rowId}/${fieldName}/${hashedName}`,
  };
}

export async function deleteFileRef(dataDir: string, ref: FileRef): Promise<void> {
  const filePath = `${dataDir}/${ref.path}`;
  try {
    await Deno.remove(filePath);
  } catch {
    // File already gone
  }
}

export async function deleteRowFiles(
  dataDir: string,
  tableName: string,
  rowId: string,
): Promise<void> {
  const dirPath = `${dataDir}/_files/${tableName}/${rowId}`;
  try {
    await Deno.remove(dirPath, { recursive: true });
  } catch {
    // Directory doesn't exist
  }
}

export async function cleanupOrphanedFiles(
  dataDir: string,
  tableName: string,
  rowId: string,
  fieldName: string,
  currentRefs: FileRef[],
): Promise<void> {
  const dirPath = `${dataDir}/_files/${tableName}/${rowId}/${fieldName}`;
  const currentPaths = new Set(currentRefs.map((r) => r.path));

  try {
    for await (const entry of Deno.readDir(dirPath)) {
      const refPath = `_files/${tableName}/${rowId}/${fieldName}/${entry.name}`;
      if (!currentPaths.has(refPath)) {
        await Deno.remove(`${dirPath}/${entry.name}`);
      }
    }
  } catch {
    // Directory doesn't exist
  }
}

// Serve a file from disk — returns Response or null
export async function serveFile(
  dataDir: string,
  urlPath: string,
): Promise<Response | null> {
  // urlPath: "/_files/users/abc123/avatar/photo.png"
  const relativePath = urlPath.replace(/^\/_files\//, "_files/");
  const filePath = `${dataDir}/${relativePath}`;

  try {
    const file = await Deno.open(filePath, { read: true });
    const stat = await file.stat();
    const mime = mimeFromExtension(filePath);

    return new Response(file.readable, {
      headers: {
        "Content-Type": mime,
        "Content-Length": String(stat.size),
        "Cache-Control": "public, max-age=31536000, immutable",
      },
    });
  } catch {
    return null;
  }
}

File: go/internal/storage/files.go
package storage

import (
	"fmt"
	"hash/fnv"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/marcisbee/flop/internal/schema"
)

// MIME type mappings.
var extToMime = map[string]string{
	".png":  "image/png",
	".jpg":  "image/jpeg",
	".jpeg": "image/jpeg",
	".gif":  "image/gif",
	".webp": "image/webp",
	".pdf":  "application/pdf",
	".json": "application/json",
	".txt":  "text/plain",
	".html": "text/html",
	".css":  "text/css",
	".js":   "application/javascript",
	".svg":  "image/svg+xml",
	".mp4":  "video/mp4",
	".mp3":  "audio/mpeg",
	".wav":  "audio/wav",
	".zip":  "application/zip",
}

// MimeFromExtension returns the MIME type for a filename.
func MimeFromExtension(filename string) string {
	ext := strings.ToLower(filepath.Ext(filename))
	if m, ok := extToMime[ext]; ok {
		return m
	}
	return "application/octet-stream"
}

// SanitizeFilename removes dangerous characters from a filename.
func SanitizeFilename(name string) string {
	name = strings.Map(func(r rune) rune {
		switch r {
		case '/', '\\', ':', '*', '?', '"', '<', '>', '|':
			return '_'
		default:
			return r
		}
	}, name)
	name = strings.ReplaceAll(name, "..", "_")
	name = strings.TrimSpace(name)
	if name == "" {
		return "unnamed"
	}
	return name
}

func hashFilename(data []byte, originalName string) string {
	h := fnv.New32a()
	h.Write(data)
	hash := strconv.FormatUint(uint64(h.Sum32()), 36)
	ext := filepath.Ext(strings.ToLower(originalName))
	return hash + ext
}

// StoreFile saves a file to the data directory and returns a FileRef.
func StoreFile(dataDir, tableName, rowID, fieldName, filename string, data []byte, mime string) (*schema.FileRef, error) {
	hashedName := hashFilename(data, filename)
	dirPath := filepath.Join(dataDir, "_files", tableName, rowID, fieldName)
	if err := os.MkdirAll(dirPath, 0755); err != nil {
		return nil, err
	}

	filePath := filepath.Join(dirPath, hashedName)
	if err := os.WriteFile(filePath, data, 0644); err != nil {
		return nil, err
	}

	relativePath := fmt.Sprintf("_files/%s/%s/%s/%s", tableName, rowID, fieldName, hashedName)
	return &schema.FileRef{
		Path: relativePath,
		Name: filename,
		Size: int64(len(data)),
		Mime: mime,
		URL:  fmt.Sprintf("/api/files/%s/%s/%s/%s", tableName, rowID, fieldName, hashedName),
	}, nil
}

// DeleteFileRef removes a file from disk.
func DeleteFileRef(dataDir string, ref *schema.FileRef) error {
	filePath := filepath.Join(dataDir, ref.Path)
	if err := os.Remove(filePath); err != nil && !os.IsNotExist(err) {
		return err
	}
	parts := strings.Split(filepath.ToSlash(strings.TrimPrefix(ref.Path, "_files/")), "/")
	if len(parts) == 4 {
		thumbDir := filepath.Join(dataDir, "_thumbs", parts[0], parts[1], parts[2])
		pattern := filepath.Join(thumbDir, "*_"+parts[3])
		if matches, err := filepath.Glob(pattern); err == nil {
			for _, match := range matches {
				_ = os.Remove(match)
			}
		}
	}
	return nil
}

// DeleteRowFiles removes the entire row's file directory.
func DeleteRowFiles(dataDir, tableName, rowID string) error {
	dirPath := filepath.Join(dataDir, "_files", tableName, rowID)
	return os.RemoveAll(dirPath)
}

// ValidateMimeType checks if a declared MIME type is in the allowed list.
func ValidateMimeType(declared string, allowed []string) bool {
	if len(allowed) == 0 {
		return true
	}
	for _, a := range allowed {
		if a == declared {
			return true
		}
		if strings.HasSuffix(a, "/*") && strings.HasPrefix(declared, a[:len(a)-1]) {
			return true
		}
	}
	return false
}
